### DISCUSSION TITLE
Specifying mixedTrans Effects

### DISCUSSION THREAD

**Discussion by non87** [stocnet/goldfish#47] [Q&A] [answered]
Hello again. I am trying to estimate a model with `mixedTrans` effect, but I apparently cannot do this. I suspect this may be a bug, but more on that below.

I can reproduce the error I am encountering with the Social Evolution data. Here it is:

```
data("Social_Evolution")
actors <- defineNodes(nodes = actors)
calls_delayed <- calls
calls_delayed$time <- calls$time + 3600
call.Network <- defineNetwork(nodes = actors, directed = TRUE)
call.Network <- linkEvents(x = call.Network, changeEvent = calls,
                           nodes = actors)
call.Network.delay <- defineNetwork(nodes = actors, directed = TRUE)
call.Network.delay <- linkEvents(x = call.Network.delay, changeEvent = calls_delayed,
                           nodes = actors)
calls <- defineDependentEvents(events = calls, nodes = actors,
                               defaultNetwork = call.Network)

estimate(calls ~ mixedTrans(call.Network.delay, window=259200) + indeg(call.Network),
         model = "DyNAM", subModel = "choice")
```

This code outputs `Error: Effect ‘mixedTrans’ (1)  cannot initialized with objects call.Network.delay_259200 missing value where TRUE/FALSE needed`.  I understand the `mixedTrans` effect in the model above is not very meaningful, but this is simply some code to reproduce the error.

I have followed the error a little in debug mode.  It all comes down to this anonymous function, where the `network` object passed is probably not what the function is expecting:

```
function (effectFun, network, window, n1, n2) 
{
  params <- formals(effectFun)
  isTwoMode <- eval(params[["isTwoMode"]])
  funApply <- eval(params[["transformFun"]])
  network2 <- sign(network[[2]])
  network1 <- sign(network[[1]])
  if (ncol(network1) != nrow(network2) || nrow(network1) != 
    n1 || ncol(network2) != n2) 
    stop("Non conformable dimensions.\n\tnetwork 1: ", paste(dim(network1), 
      collapse = ", "), "\n\tnetwork 2: ", paste(dim(network1), 
      collapse = ", "), "\n\tdependet network: ", n1, 
      ", ", n2, "\n\trows of network 1 and cols of network 2 must be the same as the correspondent in dependent network,", 
      "cols of network 1 must be the same as rows of network2")
  if ((!is.null(window) && !is.infinite(window)) || all(network1 == 
    0) || all(network2 == 0)) {
    return(list(cache = matrix(0, nrow = n1, ncol = n2), 
      stat = matrix(forceAndCall(1, funApply, 0), nrow = n1, 
        ncol = n2)))
  }
  cache <- unname(network1 %*% network2)
  return(list(cache = cache, stat = forceAndCall(1, funApply, 
    cache)))
}
```

I would appreciate any help. I honestly suspect this may be a bug in the aforementioned function because I cannot see anything wrong in the network I pass to this effect. However, I am hesitant reporting it since I may just be missing something. Thanks!

**Comment by auzaheta** [comment]
Hi @non87,

Maybe an additional description of what you expect to find using this effect would be helpful (or maybe what you understand it does). 
The mixed effects are defined to consider two networks. The documentation shows the following call:

```r
## Mixed networks
mixedTrans(network = list(network1, network2), window = Inf,
           ignoreRep = FALSE, transformFun = identity)
mixedCycle(network = list(network1, network2), window = Inf,
           ignoreRep = FALSE, transformFun = identity)
mixedClSender(network = list(network1, network2), window = Inf,
              ignoreRep = FALSE, transformFun = identity)
mixedClReceiver(network = list(network1, network2), window = Inf,
                ignoreRep = FALSE, transformFun = identity)
```

I suspect you want something like

```r
estimate(calls ~ mixedTrans(list(call.Network.delay, call.network), window=259200) + indeg(call.Network),
         model = "DyNAM", subModel = "choice")
```

The parsing cannot feed the dependent network like other effects that make it the default value. The user must specify the two networks.
However, I don't know what it means or how to interpret the parameter because it's using the same network and future info. But, it can also be just an artifact of the reproducible example you have here.

 One note, for directed networks the order matters.  `list(net1, net2)` is different from `list(net2, net1)`. The interpretation would change also. 

I'm happy to follow up on the discussion. Thanks a lot for posting these questions. I think they would help other users.

**Reply by non87** [reply]
Hi @auzaheta, thanks for the very quick answer.

I guess my question is about the syntax, which I find a little puzzling.   I find a little strange that you have to pass two networks. I guess one of them must be the dependent network. Otherwise, what are you even estimating as an effect in your choice model for the dependent network? 
From the syntax, it looks like may even pass two covariate networks to the effect...which I imagine will raise an error eventually. 

This makes me think I am not really understanding the effect. Could you walk me through the definition of the effect? I'll write down my current understanding. 
From the documentation:

> [mixedTrans] is the tendency to create an event i->j if it closes a two-paths with events (i->k) in network1 and (k->j) in network2 in the past events.

Now, the way I think about this is the following. Let's consider the possible event connecting i->j in the dependent network (indicated with x_{i->j}, which will equal one in the calculation of the choice model).  Let's consider actor k, different from actors i and j. Actor k may be connected with i and j in the covariate and/or dependent network -- I indicate these connections with the variables y_{i->k} and y_{k->j} (covariate network) and x_{i->k}, x_{k->j} (dependent network). All of these variables are either 1 or 0. The model tries to estimate the effect of the statistics x_{i->j} * x_{i->k} * y_{k->j}.

Could you please confirm my understanding makes sense?  Thanks again for your help and, more in general, for the package! 

PS Writing this, I realize that the effect I really want is x_{i->j} * y_{i->k} * y_{k->j}! I believe there is no such an effect currently, but I may just misunderstanding the syntax or the interpretation. I should point out that my network is not directed, so the arrows are arbitrary here. Thanks again for making me walk through this.

**Comment by jhollway** [comment, marked as answer]
Hi @non87, if you are interested in `x_{i->j} * y_{i->k} * y_{k→j}`, then you should be able to just pass the `y` network to `trans()`; the `x` network remains the same dependent events that you have already indicated on the left-hand side of the formula.

**Comment by non87** [comment]
@jhollway  Thanks! I completely missed this possibility. This is very helpful.

A short aside on my experience as an inexperienced end-user -- hopefully, it can help the development. In this case, the syntax really went against my intuition. 
Specifically, I cannot completely reconcile the fact that you have two different `tie` and `intertia` effects depending on you using the dependent/ covariate network for substantially the same effect -- from the documentation, `'inertia' is similar to 'tie', but defined on the network to which the dependent events relate.` Yet, you have just one effect `trans` for transitivity and you specify the network with an argument. By the way, I guess you could specify all the other closure statistics on a covariate network in a similar way (?).
I understand this discrepancy may be due to very good reasons I am not getting. However, it is still a little puzzling.  I believe it would really help if you mention the possibility (and consequences) of passing a covariate network as an argument to a closure effect in the `Goldfish Effects` help page. 

Anyway, thanks again for all the help. Truly appreciated

**Reply by jhollway** [reply]
Thanks @non87 . More documentation is on its way, and your feedback is very helpful to this end. Thanks for the feedback.

**Reply by non87** [reply]
Happy to help, thanks again for your help with my issue and for the GoldFish package more in general!
